### ISSUE TITLE
mutate does not handle error handling normally

### ISSUE THREAD

**Issue by DGURI** [awslabs/aws-mobile-appsync-sdk-js#232]
I am developing appsync and lambda in Vue.js. But there is a problem.
I want to handle the mutate error, but promise continues to return resolved.
resolved also returns null for data. However, if you check with apollo-link-error, it will return an error to the console exactly. Please understand that I am not an English-speaking person, but I am not good at English, so I do not understand it by asking a translator. So I leave a capture pattern. Please help me.

appsync.js

```javascript
import Vue from 'vue';
import { ApolloLink } from 'apollo-link';
import { setContext } from 'apollo-link-context';
import { onError } from 'apollo-link-error';
import VueApollo from 'vue-apollo';
import AWSAppSyncClient, { createAppSyncLink } from 'aws-appsync';
import { Auth } from 'aws-amplify';

import awsExports from '@/aws-exports';

Vue.use(VueApollo);

const authMiddleware = setContext((operation, previousContext) => {
  const { headers } = previousContext;
  return Auth.currentSession().then((data) => {
    const jwt = data.getIdToken().getJwtToken();
    return {
      ...previousContext,
      headers: {
        ...headers,
        jwt,
      },
    };
  });
});

const errorLink = onError(({ networkError, graphQLErrors }) => {
  if (graphQLErrors) {
    graphQLErrors.map(({ message, locations, path }) =>
      console.error(`[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`));
  }

  if (networkError) console.log(`[Network error]: ${networkError}`);
});

const AppSyncClientConfig = {
  url: awsExports.aws_appsync_graphqlEndpoint,
  region: awsExports.aws_appsync_region,
  auth: {
    type: awsExports.aws_appsync_authenticationType,
    credentials: () => Auth.currentCredentials(),
  },
};

const appSyncLink = createAppSyncLink(AppSyncClientConfig);

const link = ApolloLink.from([authMiddleware, errorLink, appSyncLink]);

const AppSyncClientOptions = {
  link,
  defaultOptions: {
    query: {
      fetchPolicy: 'network-only',
      errorPolicy: 'all ',
    },
    mutate: {
      fetchPolicy: 'network-only',
      errorPolicy: 'all ',
    },
  },
};

export const AppSyncClient = new AWSAppSyncClient({}, AppSyncClientOptions);

export default new VueApollo({
  defaultClient: AppSyncClient,
});

```

mutate code
```javascript
  onViewPost(id) {
      const { category } = this;
      const variables = {
        id,
      };
      let post = this.posts.items.find(item => item.id === id);

      if (!post) {
        post = {
          id,
          created: moment().toISOString(),
          updated: moment().toISOString(),
          title: '',
          content: '',
          category: '',
          viewer_cnt: 0,
          comment_cnt: 0,
          nickname: '',
        };
      }
      return this.$apollo.mutate({
        mutation: ViewPosts,
        variables,
        update: (store, { data: { viewPosts } }) => {
          const data = store.readQuery({ query: ListPosts, variables: { category } });
          const index = data.listPosts.items.findIndex(item => item.id === post.id);
          data.listPosts.items[index] = viewPosts;
          store.writeQuery({ query: ListPosts, data, variables: { category } });
        },
        optimisticResponse: {
          __typename: 'Mutation',
          viewPosts: {
            __typename: 'Posts',
            ...post,
          },
        },
      }).catch((err) => {
        console.error(err);
      });
    },
```
chrome console
![Image](https://ucdfb064a60e5d142357b2d7c54e.previews.dropboxusercontent.com/p/thumb/AAJe139WZ0E2piuU4HbWJ-CQ4NrhcwEz2lA59QHbiU2KhGggwXxGAu5mzEcA0zhZNRo1jU6J0uetbdyNnvKIewT_DsWht7bqHNkHtMWwmZ8Rou9wdt3nn5YSpVvSoaKTgbmzgVjuCj7y2hAz0JOe1pCqkkMF0YL15VlEy9ZhzxLUz-wF_Q3_OP-WmqL99LxCbWOxE42IxNqdJnYwKJiMjPLobaaFrqZ9t6gj8B8XUagsFQ/p.png?size=2048x1536&size_mode=3)

**Comment by raducretu** [comment]
+1 also return null but writes in DynamoDB

**Comment by manueliglesias** [comment]
Hi

Can you try with an experimental build? I've been working on improving the error handling

`npm i aws-appsync@deepdish-next aws-appsync-react@deepdish-next`

**Comment by bernhardt1** [comment]
@manueliglesias I tried your updated code. I've also tried the recent update to react-apollo @2.2.3 that was meant to fix the issue:
https://github.com/apollographql/react-apollo/releases/tag/v2.2.3

I have the same issue with my error not returning the dynamoDB data as expected. my conditional update fails (as expected) and in Cloudwatch I am able to see that Appsync returns the existing Dynamodb record along with the conditionalCheckFailedException.

The problem pops up on the front end. The client side returns only the conditional error:
[Error: GraphQL error: The conditional request failed (Service: AmazonDynamoDBv2; Status Code: 400; Error Code: ConditionalCheckFailedException; Request ID: 7XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXO5AEMVJF66Q9ASUAAJG)

Thanks for the work you do on this repo, I've been saved by your contributions a few times already!
